### ISSUE TITLE
global-buffer-overflow in ManDocVisitor::operator()(DocHtmlList const&) mandocvisitor.cpp:640

### ISSUE THREAD

**Issue by asarubbo** [doxygen/doxygen#10258]
While trying to build [efl](https://download.enlightenment.org/rel/libs/efl/efl-1.26.3.tar.xz) on the [gentoo tinderbox](https://blogs.gentoo.org/ago/2020/07/04/gentoo-tinderbox/) I got a build failure reported at [bug 912986](https://bugs.gentoo.org/912986).
After a bit of investigation about the `segmentation fault` reported by doxygen I discovered that it looks like a security bug.
I'm attaching the ASan stacktrace after I have recompiled doxygen-1.9.7 with Address Sanitizer:
[asan.log.txt](https://github.com/doxygen/doxygen/files/12436920/asan.log.txt)

The doxy file used are already public, so is useless attach it here.
If I can help further please let me know.

**Comment by albert-github** [comment]
> The doxy file used are already public, so is useless attach it here.

Where is it? Looks like it is in the attached `efl` file as `Doxyfile.in`

I looked at the mentioned `asan.log.txt` file a the stack trace shows that the problem is probably in generating the  man output type. Looking at line 640 of mandocvisitor.cpp I see:
```
  man_listItemInfo[m_indent].number = 1;
```
and a little bit earlier
```
  m_indent+=2;
```
and the `man_listItemInfo` is defined as:
```
ManListItemInfo man_listItemInfo[man_maxIndentLevels];
```
with
```
const int man_maxIndentLevels = 13;

extern ManListItemInfo man_listItemInfo[man_maxIndentLevels];
```
This problem looks like to be confirmed by:
```
0x55a5685dbb70 is located 8 bytes after global variable 'man_listItemInfo' defined in '/var/tmp/portage/app-doc/doxygen-1.9.7-r1/work/doxygen-1.9.7/src/mandocvisitor.cpp:31:17' (0x55a5685dbb00) of size 104
```
as well (104 = 13 * (4 +4), i.e. 4 byes for the int, 1 byte for the char and 3 padding bytes for each "array" element of `man_listItemInfo`).


It looks a bit like that we have a special / deeply nested HTML construct in the source.

Looking at the build log with the bugzilla issue (this log should be attached to this issue as well for consistency and persistence) I see not indication of a crash, I see though:
```
/var/tmp/portage/dev-libs/efl-1.26.3-r2/work/efl-1.26.3/doc/main.dox:32: warning: unable to resolve reference to 'elementary_main' for \ref command
/var/tmp/portage/dev-libs/efl-1.26.3-r2/work/efl-1.26.3/doc/main.dox:39: warning: HTML tag ('<a/>') may not use the 'empty tag' XHTML syntax.
```
which comes from the line:
```
 see the full <a href="modules.html" />module listing</a>.
```
that actually should read:
```
 see the full <a href="modules.html">module listing</a>.
```
 
As mentioned I didn't see a crash in the log file so no direct idea where the crash occurs,
- do you have a, doxygen,  log file from the crashing run? only the last 10 lines of this doxygen log file should be a good start indication of the place of the problem
- what happens when you fix the problem in `main.dox`?
- what happens when you use in the `INPUT` setting in the `Doxyfile.in` just the `dox` files?
- what happens when you use the current master?

To be sure we can reproduce it in one way or another:
- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

**Comment by albert-github** [comment]
In the mean time I identified the offending file (seen from the doc directory):
```
../src/bin/edje/edje_cc_handlers.c
```

**Comment by doxygen** [comment]
Looks like a specific issue in the man page output. The other output formats have similar constructs but are protected from overflowing the array, like so:

```cpp
int RTFDocVisitor::indentLevel() const
{
  return std::min(m_indentLevel,maxIndentLevels-1);
}
```

So, a workaround is to disable man page output (GENERATE_MAN=NO).

@asarubbo Note that you mention this might be a security issue, but I disagree. Doxygen is not a web application, nor is it executed with elevated privileges, so a buffer overflow can crash the application and is a bug, but does not impact security.

**Comment by albert-github** [comment]
@ doxygen
I already found the problem:
- the buffer has to be enlarged (as it uses 2 elements at a time)
- a test, like with the other output generators, has to be included.

I create a proposed PR when finishing the current one (eoither today or tomorrow).

**Comment by doxygen** [comment]
@albert-github I already started working on a fix (see referenced commit). 
@asarubbo Please verify if this fixes the problem for you. Do not close the issue, this will be done automatically when the next official release is out.

**Comment by asarubbo** [comment]
> Looking at the build log with the bugzilla issue (this log should be attached to this issue as well for consistency and persistence) I see not indication of a crash, I see though:

Please see on line 22549 of the attached log:
> 29 Segmentation fault      | doxygen -



> * do you have a, doxygen,  log file from the crashing run? only the last 10 lines of this doxygen log file should be a good start indication of the place of the problem

Here is what I have:

```
Generating namespace index...
Generating namespace member index...
Generating concept index...
Generating annotated compound index...
Generating alphabetical compound index...
Generating hierarchical class index...
Generating member index...
Generating file index...
Generating file member index...
Generating example index...
finalizing index lists...
writing tag file...
Running plantuml with JAVA...
type lookup cache used 53831/65536 hits=1306287 misses=57834
symbol lookup cache used 65536/65536 hits=1287709 misses=175028
Note: based on cache misses the ideal setting for LOOKUP_CACHE_SIZE is 2 at the cost of higher memory usage.
finished...
Segmentation fault
```

> * Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

I think that the best idea is let you able to reproduce the issue by yourself, then it will be easy do whatever :)

If you have a working docker machine, please follow these steps:

1) docker run -it -d --privileged --name mymachine gentoo/stage3
2) docker exec -it mymachine /bin/bash

Then in the container:
1) emerge --sync
2) echo "app-doc/doxygen debug" >> /etc/portage/package.use/test
3) echo "dev-libs/efl doc" >> /etc/portage/package.use/test
4) echo "media-libs/libglvnd X" >> /etc/portage/package.use/test
5) echo "media-libs/harfbuzz icu" >> /etc/portage/package.use/test
6) emerge -v dev-libs/efl

emerge will take a while because you need to compile all needed DEPEND.

In any case I'm attaching a tarball where I can reproduce the crash. Just run `doxygen` into the doc directory.
Please note that this is a tar.xz file renamed as zip because of the github policy about the attachments.
[efl-1.26.3.xz.zip](https://github.com/doxygen/doxygen/files/12438063/efl-1.26.3.xz.zip)


Please also be aware that we apply [these patches](https://github.com/gentoo/gentoo/tree/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files) over the upstream tarball.

**Comment by albert-github** [comment]
@asarubbo thanks for the effort, looks like multiple people have been working hard on the issue, please verify the changes as made by @doxygen in [13a2171](https://github.com/doxygen/doxygen/commit/13a2171757fffff32940927f3fffa9fcfbc8d472)


@doxygen, I quickly walked through the code and it looks OK to me (a bit different from what I had in mind, but probably more consistent with the other output generators.).

**Comment by asarubbo** [comment]
> @asarubbo Note that you mention this might be a security issue, but I disagree. Doxygen is not a web application, nor is it executed with elevated privileges, so a buffer overflow can crash the application and is a bug, but does not impact security.

I don't agree here.
While an application parses an external untrusted input, and the input can abuse the program, then is considered a security issue when:
1) There is a read or write memory issue in a library
2) There is a crash in a library (it does not matter what type of crash)
3) There is a write issue in a command line tool

Is not considered a security issue whatever type of crash (except memory write as stated above) in a command line tool.

Here the issue is in `mandocvisitor.cpp` that is part of libdoxymain.a and this is the first case above.

These are the [mitre rules](https://www.openwall.com/lists/oss-security/2016/09/09/11) that were stated on oss-security when I was pretty active on fuzzing.

**Comment by albert-github** [comment]
@doxygen 
With my, a bit older compiler, I get the errors:
```
...\doxygen\src\mandocvisitor.h(139): error C2079: 'ManDocVisitor::m_listItemInfo' uses undefined class 'std::array<ManDocVisitor::ManListItemInfo,13>'
...\doxygen\src\mandocvisitor.cpp(641): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(642): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(647): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(653): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(679): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(682): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(685): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(688): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(691): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(694): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(697): error C2109: subscript requires array or pointer type
...\doxygen\src\mandocvisitor.cpp(700): error C2109: subscript requires array or pointer type
..\doxygen\src\mandocvisitor.cpp(704): error C2109: subscript requires array or pointer type
```

adding 
```
#include <array>
```

solves this issue.

**Comment by doxygen** [comment]
> I don't agree here. While an application parses an external untrusted input, and the input can abuse the program, then is considered a security issue when:
>
>  [list omitted]
> 
> Here the issue is in `mandocvisitor.cpp` that is part of libdoxymain.a and this is the first case above.

I can somewhat understand where these 'rules' come from, but the library you mention is an integral, and private part of doxygen, not something any other application will ever link to. So please see the static library as a bundle of `.o` file that could have just as well been directly compiled into an executable.

**Comment by albert-github** [comment]
@asarubbo 

Regarding the:
> Please also be aware that we apply [these patches](https://github.com/gentoo/gentoo/tree/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files) over the upstream tarball.

in https://github.com/doxygen/doxygen/issues/10258#issuecomment-1693128710

I see that there are 4 patches:
- [doxygen-1.9.1-do_not_force_libcxx.patch](https://github.com/gentoo/gentoo/blob/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files/doxygen-1.9.1-do_not_force_libcxx.patch)
   Wouldn't it be better do incorporate this in the build procedure on the cmake line like `-Duse_libc++=OFF`?
- [doxygen-1.9.1-ignore-bad-encoding.patch](https://github.com/gentoo/gentoo/blob/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files/doxygen-1.9.1-ignore-bad-encoding.patch)
  Which files did have a problem? Has this been reported upstream,  as the patch looks simple enough and probably not necessary anymore for the current doxygen version (when I remember correctly some changes were made to the files.
- [doxygen-1.9.4-link_with_pthread.patch](https://github.com/gentoo/gentoo/blob/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files/doxygen-1.9.4-link_with_pthread.patch)
   Is this something special for the gentoo OS?
- [doxygen-1.9.7-musl-1.2.4.patch](https://github.com/gentoo/gentoo/blob/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files/doxygen-1.9.7-musl-1.2.4.patch)
   @doxygen looks like this problem is fixed around 7 month ago as so in  the 1.12.0 version, maybe doxygen should update.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.8.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).

**Comment by thesamesam** [comment]
> @asarubbo
> 
> Regarding the:
> 
> > Please also be aware that we apply [these patches](https://github.com/gentoo/gentoo/tree/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files) over the upstream tarball.
> 
> in [#10258 (comment)](https://github.com/doxygen/doxygen/issues/10258#issuecomment-1693128710)
> 

Thank you for your feedback!

> I see that there are 4 patches:
> 
>     * [doxygen-1.9.1-do_not_force_libcxx.patch](https://github.com/gentoo/gentoo/blob/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files/doxygen-1.9.1-do_not_force_libcxx.patch)
>       Wouldn't it be better do incorporate this in the build procedure on the cmake line like `-Duse_libc++=OFF`?
> 

Yes - done.

>     * [doxygen-1.9.1-ignore-bad-encoding.patch](https://github.com/gentoo/gentoo/blob/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files/doxygen-1.9.1-ignore-bad-encoding.patch)
>       Which files did have a problem? Has this been reported upstream,  as the patch looks simple enough and probably not necessary anymore for the current doxygen version (when I remember correctly some changes were made to the files.
> 

This was quite a while ago when I was still quite new and I wasn't confident enough with investigating. I suspect this issue is indeed gone, but I'll report it if it's still there.

>     * [doxygen-1.9.4-link_with_pthread.patch](https://github.com/gentoo/gentoo/blob/8916d7f05e64aeb62e5adb4b7cc206d8829ad72e/app-doc/doxygen/files/doxygen-1.9.4-link_with_pthread.patch)
>       Is this something special for the gentoo OS?

I'd have to go digging to see why this one got added, it predates me.

Appreciate the help!
